lib/src/main/scala/spinal/lib/tools/ModuleAnalyzer.scala

```diff
@@ -0,0 +1,166 @@
+        }
+      }
+    }
+    ret.result()
+  }
+
+  /** Filter the clock domains
+    * @param filter - the predicate to filter the clock domains
+    * @return set of clock domain
+    */
+  def getClocks(filter: ClockDomain => Boolean): mutable.LinkedHashSet[ClockDomain] = allClocks.filter(filter)
+
+  /** Get all the registers inside the module
+    * @return set of register of base type
+    */
+  def allRegisters: mutable.LinkedHashSet[BaseType] = {
+    val ret = mutable.LinkedHashSet.newBuilder[BaseType]
+    module.dslBody.walkDeclarations {
+      case ds: BaseType if ds.isReg =>
+        ret += ds
+      case _ =>
+    }
+    ret.result()
+  }
+
+  /** Filter the registers
+    * @param filter  - the predicate to filter the clock domains
+    * @return set of filtered registers
+    */
+  def getRegisters(filter: BaseType => Boolean): mutable.LinkedHashSet[BaseType] = allRegisters.filter(filter)
+
+  /** Get the submodule instances that meet the condition
+    * @param cond - the condition that instance should meet.
+    * @return - set of sub module instances
+    */
+  def getCells(cond: Module => Boolean): mutable.LinkedHashSet[Module] = {
+    val ret = mutable.LinkedHashSet.newBuilder[Module]
+    module.walkComponents { m =>
+      if (cond(m)) ret += m
+    }
+    ret.result()
+  }
+
+  /** Get the sub-blackbox instances that meet the condition
+    * @param cond - the condition that instance should meet.
+    * @return - set of sub blackbox instances
+    */
+  def getLibCells(cond: BlackBox => Boolean): mutable.LinkedHashSet[BlackBox] = {
+    val ret = mutable.LinkedHashSet.newBuilder[BlackBox]
+    module.walkComponents {
+      case bb: BlackBox if cond(bb) =>
+        ret += bb
+      case _ =>
+    }
+    ret.result()
+  }
+
+  /** Get the wire/net inside the module
+    * @param cond the filtering condition
+    * @return set of the base type nets
+    */
+  def getNets(cond: BaseType => Boolean): mutable.LinkedHashSet[BaseType] = {
+    val ret = mutable.LinkedHashSet.newBuilder[BaseType]
+    module.walkComponents { m =>
+      m.dslBody.walkDeclarations {
+        case ds: BaseType if cond(ds) =>
+          ret += ds
+        case _ =>
+      }
+    }
+    ret.result()
+  }
+
+  /** Get pins of the sub-module instance inside the module.
+    * @param cond the filtering condition
+    * @return set of the pin type
+    */
+  def getPins(cond: BaseType => Boolean): mutable.LinkedHashSet[BaseType] = {
+    val ret = mutable.LinkedHashSet.newBuilder[BaseType]
+    module.walkComponents { m =>
+      m.dslBody.walkDeclarations {
+        case ds: BaseType if (ds.isInputOrInOut || ds.isOutput) && cond(ds) =>
+          ret += ds
+        case _ =>
+      }
+    }
+    ret.result()
+  }
+
+  /** Get pins of the sub-blackbox instance inside the module.
+    * @param cond the filtering condition
+    * @return set of the pin type
+    */
+  def getLibPins(cond: BaseType => Boolean): mutable.LinkedHashSet[BaseType] = {
```

Comment:
I don't understand `getLibPins` vs `getPins`

Reply:
`getLibPins` returns only the blackbox instances' pins, while `getPins` returns all instances' pins.

Reply:
I do not understand the relationship between "lib" and "BlackBox", maybe because I don't know `sdc` enough XD

Reply:
"lib" here means a library containing IP blocks/macros. Those things in a "lib" are represented as `BlackBox` in SpinalHDL.

Reply:
Okay I better understand, thanks!